### DISCUSSION TITLE
Improve the handling of error reports (`var/report`)

### DISCUSSION THREAD

**Discussion by loekvangool** [OpenMage/magento-lts#3994] [Ideas]
When OM runs into an exception very early in the execution of the code, before `Mage::log()` or `Mage::logException()` can take care of it, the following code comes into play: https://github.com/OpenMage/magento-lts/tree/main/errors. It basically writes a dump of the error and the trace to `var/report/` in a extension-less file.

Several things that I'm not comfortable with:
1. The error report can be emailed or logged, but not both
2. The error report via email contains more information (like IP address, customer details) that are not available in the logged error
3. The error report log has no extension, is written using `serialize()` (maybe JSON is better), creates one file per error, and is placed in a directory that is not self-explanatory so it might be overlooked.
4. There is no cleanup of these files, there could be millions.
5. The email handler is using the PHP's `email()` (not `Zend_Mail`), does not care about SMTP plugins etc. and I can understand that, as this handler is separated from OM and cannot rely on those libs. However, there is no throttling hence it could seriously create problems if hundreds of these emails are sent every minute in the case of, for example, an unreachable database.

My suggestion would be to start writing these messages to a new NDJSON file in `var/log/`, along with all the metadata that the email has (maybe more). But I would appreciate any thoughts from you all!

**Comment by luigifab** [comment]
Sentry does all this, more, and better.
For example you can configure some alerts, if the same error occurs more than 10 times in one hour, you can receive an email, a slack alert...

**Comment by loekvangool** [comment]
Yes and I have another APM that catches it as well. But I'm not sure what you mean with that. The out of the box product takes a weird direction here, using a parallel logging system that some users will miss and has an email system that is likely to shutdown a company's email system if there is ever an error on the site, and also it's full of `@` statements and hardcoded paths and file permissions. APM solutions are optional and not great to plug a hole in the product.
